From a002087260fdf02726720e6452ab735d90567401 Mon Sep 17 00:00:00 2001
From: Lukas Czerner <lczerner@redhat.com>
Date: Thu, 23 Oct 2014 09:54:16 -0400
Subject: [fs] ext4: fix error paths in swap_inode_boot_loader()

Message-id: <1414058125-4183-103-git-send-email-lczerner@redhat.com>
Patchwork-id: 98189
O-Subject: [RHEL 7.1 PATCH v2 102/171] ext4: fix error paths in swap_inode_boot_loader()
Bugzilla: 1150139
RH-Acked-by: Eric Sandeen <sandeen@redhat.com>

From: Zheng Liu <wenqing.lz@taobao.com>

BZ 1150139
https://bugzilla.redhat.com/show_bug.cgi?id=1150139

BREW 8142710
https://brewweb.devel.redhat.com/taskinfo?taskID=8142710

Upstream commit 30d29b119ef01776e0a301444ab24defe8d8bef3

 In swap_inode_boot_loader() we forgot to release ->i_mutex and resume
 unlocked dio for inode and inode_bl if there is an error starting the
 journal handle.  This commit fixes this issue.

 Reported-by: Ahmed Tamrawi <ahmedtamrawi@gmail.com>
 Cc: Andreas Dilger <adilger.kernel@dilger.ca>
 Cc: Dr. Tilmann Bubeck <t.bubeck@reinform.de>
 Signed-off-by: Zheng Liu <wenqing.lz@taobao.com>
 Signed-off-by: "Theodore Ts'o" <tytso@mit.edu>
 Cc: stable@vger.kernel.org  # v3.10+
 Signed-off-by: Lukas Czerner <lczerner@redhat.com>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/fs/ext4/ioctl.c b/fs/ext4/ioctl.c
index 6bea806..a2a837f 100644
--- a/fs/ext4/ioctl.c
+++ b/fs/ext4/ioctl.c
@@ -140,7 +140,7 @@ static long swap_inode_boot_loader(struct super_block *sb,
  handle = ext4_journal_start(inode_bl, EXT4_HT_MOVE_EXTENTS, 2);
  if (IS_ERR(handle)) {
   err = -EINVAL;
-  goto swap_boot_out;
+  goto journal_err_out;
  }
 
  /* Protect extent tree against block allocations via delalloc */
@@ -198,6 +198,7 @@ static long swap_inode_boot_loader(struct super_block *sb,
 
  ext4_double_up_write_data_sem(inode, inode_bl);
 
+journal_err_out:
  ext4_inode_resume_unlocked_dio(inode);
  ext4_inode_resume_unlocked_dio(inode_bl);
 
-- 
1.7.1